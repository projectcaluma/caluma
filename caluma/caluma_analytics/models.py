from django.contrib.postgres.fields import ArrayField
from django.db import models
from django.db.models.constraints import UniqueConstraint
from localized_fields.fields import LocalizedField

from caluma.caluma_core.models import SlugModel, UUIDModel

from .simple_table import BaseStartingObject


class AnalyticsTable(SlugModel):
    STARTING_OBJECT_CHOICES = BaseStartingObject.as_choices()

    meta = models.JSONField(default=dict)

    disable_visibilities = models.BooleanField(default=False)
    name = LocalizedField(blank=False, null=False, required=False)
    starting_object = models.CharField(max_length=250, choices=STARTING_OBJECT_CHOICES)

    def get_starting_object(self, info):
        return BaseStartingObject.get_object(self.starting_object, info)


class AnalyticsField(UUIDModel):
    alias = models.CharField(max_length=100)
    meta = models.JSONField(default=dict)

    data_source = models.TextField()
    table = models.ForeignKey(
        AnalyticsTable, on_delete=models.CASCADE, related_name="fields"
    )
    filters = ArrayField(models.TextField(), blank=True, null=True, default=list)

    class Meta:
        constraints = [
            UniqueConstraint(
                name="unique_data_source", fields=["table", "data_source"]
            ),
            UniqueConstraint(name="unique_alias", fields=["table", "alias"]),
        ]
