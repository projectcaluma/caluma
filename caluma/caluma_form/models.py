import uuid

from django.contrib.postgres.fields import ArrayField, JSONField
from django.contrib.postgres.indexes import GinIndex
from django.db import models, transaction
from localized_fields.fields import LocalizedField, LocalizedTextField
from simple_history.models import HistoricalRecords

from ..caluma_core import models as core_models
from .storage_clients import client


class Form(core_models.SlugModel):
    name = LocalizedField(blank=False, null=False, required=False)
    description = LocalizedField(blank=True, null=True, required=False)
    meta = JSONField(default=dict)
    is_published = models.BooleanField(default=False)
    is_archived = models.BooleanField(default=False)
    questions = models.ManyToManyField(
        "Question", through="FormQuestion", related_name="forms"
    )
    source = models.ForeignKey(
        "self",
        blank=True,
        null=True,
        help_text="Reference this form has been copied from",
        related_name="+",
        on_delete=models.SET_NULL,
    )

    class Meta:
        indexes = [GinIndex(fields=["meta"])]


class FormQuestion(core_models.NaturalKeyModel):
    form = models.ForeignKey("Form", on_delete=models.CASCADE)
    question = models.ForeignKey("Question", on_delete=models.CASCADE)
    sort = models.PositiveIntegerField(editable=False, db_index=True, default=0)

    def natural_key(self):
        return f"{self.form_id}.{self.question_id}"

    class Meta:
        ordering = ["-sort"]
        unique_together = ("form", "question")


class Question(core_models.SlugModel):
    TYPE_MULTIPLE_CHOICE = "multiple_choice"
    TYPE_INTEGER = "integer"
    TYPE_FLOAT = "float"
    TYPE_DATE = "date"
    TYPE_CHOICE = "choice"
    TYPE_TEXTAREA = "textarea"
    TYPE_TEXT = "text"
    TYPE_TABLE = "table"
    TYPE_FORM = "form"
    TYPE_FILE = "file"
    TYPE_DYNAMIC_CHOICE = "dynamic_choice"
    TYPE_DYNAMIC_MULTIPLE_CHOICE = "dynamic_multiple_choice"
    TYPE_STATIC = "static"
    TYPE_CALCULATED_FLOAT = "calculated_float"

    TYPE_CHOICES = (
        TYPE_MULTIPLE_CHOICE,
        TYPE_INTEGER,
        TYPE_FLOAT,
        TYPE_DATE,
        TYPE_CHOICE,
        TYPE_TEXTAREA,
        TYPE_TEXT,
        TYPE_TABLE,
        TYPE_FORM,
        TYPE_FILE,
        TYPE_DYNAMIC_CHOICE,
        TYPE_DYNAMIC_MULTIPLE_CHOICE,
        TYPE_STATIC,
        TYPE_CALCULATED_FLOAT,
    )
    TYPE_CHOICES_TUPLE = ((type_choice, type_choice) for type_choice in TYPE_CHOICES)

    label = LocalizedField(blank=False, null=False, required=False)
    type = models.CharField(choices=TYPE_CHOICES_TUPLE, max_length=23)
    is_required = models.TextField(default="false")
    is_hidden = models.TextField(default="false")
    is_archived = models.BooleanField(default=False)
    placeholder = LocalizedField(blank=True, null=True, required=False)
    info_text = LocalizedField(blank=True, null=True, required=False)
    static_content = LocalizedTextField(blank=True, null=True, required=False)
    configuration = JSONField(default=dict)
    meta = JSONField(default=dict)
    data_source = models.CharField(max_length=255, blank=True, null=True)
    options = models.ManyToManyField(
        "Option", through="QuestionOption", related_name="questions"
    )
    row_form = models.ForeignKey(
        Form,
        blank=True,
        null=True,
        related_name="+",
        help_text="Form that represents rows of a TableQuestion",
        on_delete=models.PROTECT,
    )
    sub_form = models.ForeignKey(
        Form,
        blank=True,
        null=True,
        related_name="+",
        help_text="Form referenced in a FormQuestion",
        on_delete=models.PROTECT,
    )

    source = models.ForeignKey(
        "self",
        blank=True,
        null=True,
        related_name="+",
        help_text="Reference this question has been copied from",
        on_delete=models.SET_NULL,
    )
    format_validators = ArrayField(
        models.CharField(max_length=255), blank=True, default=list
    )

    default_answer = models.OneToOneField(
        "Answer", on_delete=models.SET_NULL, related_name="+", null=True, blank=True
    )

    calc_expression = models.TextField(null=True, blank=True)
    calc_dependents = ArrayField(
        models.CharField(max_length=255, blank=True), default=list
    )

    @property
    def min_length(self):
        return self.configuration.get("min_length")

    @min_length.setter
    def min_length(self, value):
        self.configuration["min_length"] = value

    @property
    def max_length(self):
        return self.configuration.get("max_length")

    @max_length.setter
    def max_length(self, value):
        self.configuration["max_length"] = value

    @property
    def max_value(self):
        return self.configuration.get("max_value")

    @max_value.setter
    def max_value(self, value):
        self.configuration["max_value"] = value

    @property
    def min_value(self):
        return self.configuration.get("min_value")

    @min_value.setter
    def min_value(self, value):
        self.configuration["min_value"] = value

    def empty_value(self):
        """Return empty value for this question type."""

        empties = {
            Question.TYPE_MULTIPLE_CHOICE: [],
            Question.TYPE_TABLE: [],
            Question.TYPE_DYNAMIC_MULTIPLE_CHOICE: [],
        }
        return empties.get(self.type, None)

    def __repr__(self):
        base = super().__repr__()
        return base[:-1] + f", type={self.type})"

    class Meta:
        indexes = [GinIndex(fields=["meta"])]


class QuestionOption(core_models.NaturalKeyModel):
    question = models.ForeignKey("Question", on_delete=models.CASCADE)
    option = models.ForeignKey("Option", on_delete=models.CASCADE)
    sort = models.PositiveIntegerField(editable=False, db_index=True, default=0)

    def natural_key(self):
        return f"{self.question_id}.{self.option_id}"

    class Meta:
        ordering = ["-sort"]
        unique_together = ("option", "question")


class Option(core_models.SlugModel):
    label = LocalizedField(blank=False, null=False, required=False)
    is_archived = models.BooleanField(default=False)
    meta = JSONField(default=dict)
    source = models.ForeignKey(
        "self",
        blank=True,
        null=True,
        related_name="+",
        help_text="Reference this option has been copied from",
        on_delete=models.SET_NULL,
    )

    class Meta:
        indexes = [GinIndex(fields=["meta"])]


class DocumentManager(models.Manager):
    @transaction.atomic
    def create_document_for_task(self, task, user):
        """Create a document for a given task."""
        if task.form_id is not None:
            # import domain logic here, in order to avoid recursive import error
            from .domain_logic import SaveDocumentLogic

            return SaveDocumentLogic.create({"form": task.form}, user=user)


class Document(core_models.UUIDModel):
    objects = DocumentManager()

    family = models.ForeignKey(
        "self",
        help_text="Family id which document belongs too.",
        null=True,
        on_delete=models.CASCADE,
        related_name="+",
    )
    form = models.ForeignKey(
        "caluma_form.Form", on_delete=models.DO_NOTHING, related_name="documents"
    )
    source = models.ForeignKey(
        "self",
        blank=True,
        null=True,
        help_text="Reference this document has been copied from",
        related_name="copies",
        on_delete=models.SET_NULL,
    )
    meta = JSONField(default=dict)

    def set_family(self, root_doc):
        """Set the family to the given root_doc.

        Apply the same to all row documents within this document,
        recursively.
        """
        if self.family != root_doc:
            self.family = root_doc
            self.save()

        table_answers = Answer.objects.filter(
            document=self,
            # TODO this is not REALLY required, as only table answers
            # have child documents anyhow. Leaving it out would avoid
            # a JOIN to the question table
            question__type=Question.TYPE_TABLE,
        )

        child_documents = AnswerDocument.objects.filter(
            answer__in=table_answers
        ).select_related("document")

        for answer_document in child_documents:
            answer_document.document.set_family(root_doc)

    def copy(self, family=None, user=None):
        """Create a copy including all its answers."""

        new_document = type(self).objects.create(
            form=self.form,
            meta=dict(self.meta),
            source=self,
            family=family,
            created_by_user=user.username if user else None,
            created_by_group=user.group if user else None,
        )
        if not family:
            family = new_document

        # copy answers
        for source_answer in self.answers.all():
            source_answer.copy(
                document_family=family, to_document=new_document, user=user
            )

        return new_document

    def __repr__(self):
        return f"Document(form={self.form!r})"

    class Meta:
        indexes = [GinIndex(fields=["meta"])]


class QuestionTypeHistoricalModel(models.Model):
    """Extra fields for HistoricalAnswer."""

    history_question_type = models.CharField(
        choices=Question.TYPE_CHOICES_TUPLE, max_length=23
    )

    class Meta:
        abstract = True


class Answer(core_models.BaseModel):
    """Records an answer to a question of arbitrary type."""

    # We need to replicate the UUIDModel in order to register it as historical model.
    # Otherwise simple_history complains that the model is already registered,
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    question = models.ForeignKey(
        "caluma_form.Question", on_delete=models.DO_NOTHING, related_name="answers"
    )
    value = JSONField(null=True, blank=True)
    meta = JSONField(default=dict)
    document = models.ForeignKey(
        Document,
        on_delete=models.CASCADE,
        related_name="answers",
        null=True,
        blank=True,
    )
    documents = models.ManyToManyField(
        Document, through="AnswerDocument", related_name="+"
    )
    date = models.DateField(null=True, blank=True)
    file = models.OneToOneField(
        "File", on_delete=models.SET_NULL, null=True, blank=True
    )

    # override history to add extra fields on historical model
    history = HistoricalRecords(
        inherit=True,
        history_user_id_field=models.CharField(null=True, max_length=150),
        history_user_setter=core_models._history_user_setter,
        history_user_getter=core_models._history_user_getter,
        bases=[QuestionTypeHistoricalModel],
    )

    def delete(self, *args, **kwargs):
        if self.file:
            self.file.delete()
        super().delete(args, kwargs)

    def create_answer_documents(self, documents):
        family = getattr(self.document, "family", None)
        document_ids = [document.pk for document in documents]

        for sort, document_id in enumerate(reversed(document_ids), start=1):
            ans_doc, created = AnswerDocument.objects.get_or_create(
                answer=self, document_id=document_id, defaults={"sort": sort}
            )
            if not created and ans_doc.sort != sort:
                ans_doc.sort = sort
                ans_doc.save()
            if created:
                # Already-existing documents are already in the family,
                # so we're updating only the newly attached rows
                ans_doc.document.set_family(family)

    def unlink_unused_rows(self, docs_to_keep):
        existing = AnswerDocument.objects.filter(answer=self).exclude(
            document__in=docs_to_keep
        )
        for ans_doc in list(existing.select_related("document")):
            # Set document to be its own family
            # TODO: Can/should we delete the detached documents?
            ans_doc.document.set_family(ans_doc.document)
            ans_doc.delete()

    def copy(self, document_family=None, to_document=None, user=None):
        new_answer = type(self).objects.create(
            question=self.question,
            value=self.value,
            meta=dict(self.meta),
            date=self.date,
            document=to_document,
            created_by_user=user.username if user else None,
            created_by_group=user.group if user else None,
        )

        if self.question.type == Question.TYPE_FILE:
            new_answer.file = self.file.copy()
            new_answer.save()

        # TableAnswer: copy AnswerDocument too
        for answer_doc in AnswerDocument.objects.filter(answer=self):
            new_doc = answer_doc.document.copy(family=document_family, user=user)

            AnswerDocument.objects.create(
                answer=new_answer,
                document=new_doc,
                sort=answer_doc.sort,
                created_by_user=user.username if user else None,
                created_by_group=user.group if user else None,
            )
        return new_answer

    def __repr__(self):
        return f"Answer(document={self.document!r}, question={self.question!r}, value={self.value!r})"

    class Meta:
        # a question may only be answerd once per document
        unique_together = ("document", "question")
        indexes = [models.Index(fields=["date"]), GinIndex(fields=["meta", "value"])]


class File(core_models.UUIDModel):
    name = models.CharField(max_length=255)

    def _move_blob(self):
        # move the file on update
        # this makes sure it stays available when querying the history
        old_file = self.history.first()
        new_name = f"{old_file.pk}_{old_file.name}"
        client.move_object(self.object_name, new_name)

    def copy(self, *args, **kwargs):
        copy = File.objects.create(name=self.name)
        client.copy_object(self.object_name, copy.object_name)
        return copy

    def delete(self, *args, **kwargs):
        self._move_blob()
        super().delete(*args, **kwargs)

    def save(self, *args, **kwargs):
        if self.created_at:
            self._move_blob()
        super().save(*args, **kwargs)

    @property
    def object_name(self):
        return f"{self.pk}_{self.name}"

    @property
    def upload_url(self):
        return client.upload_url(self.object_name)

    @property
    def download_url(self):
        return client.download_url(self.object_name)

    @property
    def metadata(self):
        return client.stat_object(self.object_name).__dict__


class AnswerDocument(core_models.UUIDModel):
    answer = models.ForeignKey("Answer", on_delete=models.CASCADE)
    document = models.ForeignKey("Document", on_delete=models.CASCADE)
    sort = models.PositiveIntegerField(editable=False, db_index=True, default=0)

    class Meta:
        ordering = ["-sort"]
        unique_together = ("answer", "document")


class DynamicOption(core_models.UUIDModel):
    slug = models.CharField(max_length=255)
    label = LocalizedField(blank=False, null=False, required=False)
    document = models.ForeignKey("Document", on_delete=models.CASCADE)
    question = models.ForeignKey("Question", on_delete=models.CASCADE)

    class Meta:
        unique_together = ("slug", "document", "question")
