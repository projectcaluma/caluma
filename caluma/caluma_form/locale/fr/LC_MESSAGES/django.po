# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-09-05 12:21+0000\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n > 1);\n"

#: caluma/caluma_form/format_validators.py:58
msgid "E-mail address"
msgstr "Adresse e-mail"

#: caluma/caluma_form/format_validators.py:60
msgid "Please enter a valid e-mail address"
msgstr "Veuillez entrer une addresse e-mail valide"

#: caluma/caluma_form/format_validators.py:65
msgid "Phone number"
msgstr "Numéro de téléphone"

#: caluma/caluma_form/format_validators.py:67
msgid "Please enter a valid phone number"
msgstr "Veuillez entrer un numéro de téléphone valide"
