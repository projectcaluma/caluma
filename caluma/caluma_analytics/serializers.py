from django.core import exceptions
from django.db import transaction

from ..caluma_core import serializers
from . import models


class SaveAnalyticsTableSerializer(serializers.ModelSerializer):
    class Meta:
        model = models.AnalyticsTable
        fields = [
            "slug",
            "name",
            "starting_object",
            "created_at",
            "modified_at",
            "disable_visibilities",
            "meta",
            "created_by_user",
            "created_by_group",
            "modified_by_user",
            "modified_by_group",
        ]


class RemoveAnalyticsTableSerializer(serializers.ModelSerializer):
    slug = serializers.GlobalIDField()

    @transaction.atomic
    def update(self, instance, validated_data):
        instance.delete()
        return instance

    class Meta:
        lookup_input_kwarg = "slug"
        fields = ["slug"]
        model = models.AnalyticsTable


class SaveAnalyticsFieldSerializer(serializers.ModelSerializer):
    id = serializers.GlobalIDField(required=False)

    def validate(self, data):
        validated_data = super().validate(data)
        self._validate_uniqueness(validated_data)
        self._validate_field_exists(validated_data)
        return validated_data

    def _validate_field_exists(self, data):
        start = data["table"].get_starting_object(self.context["info"])
        data_source = data["data_source"]
        try:
            start.get_field(data_source.split("."))
        except KeyError:
            raise exceptions.ValidationError(
                {
                    "dataSource": [
                        f"Specified data source '{data_source}' is not available"
                    ]
                }
            )

    def _validate_uniqueness(self, data):
        existing_fields = data["table"].fields.all()
        if self.instance:
            existing_fields = existing_fields.exclude(pk=self.instance.pk)

        if existing_fields.filter(alias=data["alias"]).exists():
            raise exceptions.ValidationError(
                {"alias": ["Cannot use the same alias twice within the same table"]}
            )
        if existing_fields.filter(data_source=data["data_source"]).exists():
            raise exceptions.ValidationError(
                {
                    "dataSource": [
                        "Cannot use the same data source twice within the same table"
                    ]
                }
            )

    class Meta:
        model = models.AnalyticsField
        fields = [
            "id",
            "alias",
            "table",
            "data_source",
            "created_at",
            "modified_at",
            "filters",
            "meta",
            "created_by_user",
            "created_by_group",
            "modified_by_user",
            "modified_by_group",
        ]


class RemoveAnalyticsFieldSerializer(serializers.ModelSerializer):
    id = serializers.GlobalIDField()

    @transaction.atomic
    def update(self, instance, validated_data):
        instance.delete()

        return instance

    class Meta:
        fields = ["id"]
        model = models.AnalyticsField
