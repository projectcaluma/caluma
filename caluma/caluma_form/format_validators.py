import re
from collections import namedtuple

from django.conf import settings
from django.utils.module_loading import import_string
from localized_fields.value import LocalizedValue
from rest_framework.exceptions import ValidationError


class BaseFormatValidator:
    r"""Basic format validator class to be extended by any format validator implementation.

    A custom format validator class could look like this:
    ```
    >>> from caluma.caluma_form.format_validators import BaseFormatValidator
    ...
    ...
    ... class CustomFormatValidator(BaseFormatValidator):
    ...     slug = "email"
    ...     name = {"en": "E-mail", "de": "E-Mail"}
    ...     regex = r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)"
    ...     error_msg = {"en": "Not an e-mail address", "de": "Keine E-Mail adresse"}
    ```
    """

    # `name` and `error_msg` are just dicts and get cast to `LocalizedValue` when needed.
    #  This is to not break with the existing API.

    def __init__(self):
        if not all(
            [self.slug, self.regex, self.name, self.error_msg]
        ):  # pragma: no cover
            raise NotImplementedError("Missing properties!")

    def validate(self, value, document):
        if not re.match(self.regex, value):
            raise ValidationError(LocalizedValue(self.error_msg).translate())


class EMailFormatValidator(BaseFormatValidator):
    slug = "email"
    name = {"en": "E-mail address", "de": "E-Mail Adresse", "fr": "Adresse e-mail"}
    regex = r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)"
    error_msg = {
        "en": "Please enter a valid e-mail address",
        "de": "Bitte geben Sie eine gültige E-Mail Adresse ein",
        "fr": "Veuillez entrer une addresse e-mail valide",
    }


class PhoneNumberFormatValidator(BaseFormatValidator):
    slug = "phone-number"
    name = {"en": "Phone number", "de": "Telefonnummer", "fr": "Numéro de téléphone"}
    regex = r"^[\s\/\.\(\)-]*(?:\+|0|00)(?:[\s\/\.\(\)-]*\d[\s\/\.\(\)-]*){6,20}$"
    error_msg = {
        "en": "Please enter a valid phone number",
        "de": "Bitte geben Sie eine gültige Telefonnummer ein",
        "fr": "Veuillez entrer un numéro de téléphone valide",
    }


base_format_validators = [EMailFormatValidator, PhoneNumberFormatValidator]


FormatValidator = namedtuple("FormatValidator", ["slug", "name", "regex", "error_msg"])


def get_format_validators(include=None, dic=False):
    """Get all FormatValidators.

    :param include: List of FormatValidators to include
    :param dic: Should return a dict
    :return: List of FormatValidator-objects if dic False otherwise dict
    """

    format_validator_classes = [
        import_string(cls) for cls in settings.FORMAT_VALIDATOR_CLASSES
    ] + base_format_validators
    if include is not None:
        format_validator_classes = [
            fvc for fvc in format_validator_classes if fvc.slug in include
        ]
    if dic:
        return {ds.slug: ds for ds in format_validator_classes}
    return [
        FormatValidator(
            slug=ds.slug,
            name=LocalizedValue(ds.name).translate(),
            regex=ds.regex,
            error_msg=LocalizedValue(ds.error_msg).translate(),
        )
        for ds in format_validator_classes
    ]
