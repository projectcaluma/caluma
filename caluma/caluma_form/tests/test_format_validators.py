import warnings

import pytest
from django.utils import translation

from caluma.deprecation import CalumaDeprecationWarning

from ...caluma_core.tests import extract_serializer_input_fields
from ..format_validators import (
    FORMAT_VALIDATION_FAILED,
    BaseFormatValidator,
    base_format_validators,
)
from ..models import Question
from ..serializers import SaveAnswerSerializer

with warnings.catch_warnings():
    # Explicitly ignore the expected deprecation warning as we raise warnings to
    # errors while testing. We explicitly want to test that it still works with
    # the deprecated syntax.
    warnings.filterwarnings("ignore", category=CalumaDeprecationWarning)

    class MyFormatValidator(BaseFormatValidator):
        slug = "test-validator"
        name = {"en": "test name english", "de": "test name deutsch"}
        regex = r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)"
        error_msg = {"en": "Not valid", "de": "Nicht valid"}


class MyDateValidator(BaseFormatValidator):
    slug = "my-date-validator"
    name = "my date validator"
    error_msg = "%(day)s is not an even day"
    allowed_question_types = [Question.TYPE_DATE]

    @classmethod
    def is_valid(cls, value, document):
        return value.day % 2 == 0

    @classmethod
    def get_error_msg_args(cls, value, document):
        return dict(day=value.day)


def test_fetch_format_validators(snapshot, schema_executor, settings):
    settings.FORMAT_VALIDATOR_CLASSES = [
        "caluma.caluma_form.tests.test_format_validators.MyFormatValidator",
        "caluma.caluma_form.tests.test_format_validators.MyDateValidator",
    ]
    query = """
        query formatValidators {
          allFormatValidators {
            pageInfo {
              startCursor
              endCursor
            }
            totalCount
            edges {
              node {
                slug
                name
                regex
                allowedQuestionTypes
              }
            }
          }
        }
    """

    result = schema_executor(query)
    assert not result.errors
    snapshot.assert_match(result.data)

    with translation.override("de"):
        result = schema_executor(query)
        assert not result.errors
        snapshot.assert_match(result.data)

    with translation.override("fr"):
        result = schema_executor(query)
        assert not result.errors
        snapshot.assert_match(result.data)

    with translation.override("nolang"):
        result = schema_executor(query)
        assert not result.errors
        snapshot.assert_match(result.data)


@pytest.mark.parametrize("question__type", [Question.TYPE_TEXT, Question.TYPE_TEXTAREA])
@pytest.mark.parametrize(
    "question__format_validators,answer__value,expected_error",
    [
        (["email"], "some text", "Please enter a valid e-mail address"),
        (["email"], "", None),
        (["email"], "test@example.com", None),
        (["phone-number"], "some text", "Please enter a valid phone number"),
        (["phone-number"], "+411234567890", None),
        (
            ["email", "phone-number"],
            "+411234567890",
            "Please enter a valid e-mail address",
        ),
        (
            ["email", "phone-number"],
            "test@example.com",
            "Please enter a valid phone number",
        ),
    ],
)
def test_base_format_validators(
    db, snapshot, question, answer, expected_error, schema_executor
):
    query = """
        mutation SaveDocumentStringAnswer($input: SaveDocumentStringAnswerInput!) {
          saveDocumentStringAnswer(input: $input) {
            answer {
              ... on StringAnswer {
                stringValue: value
              }
            }
            clientMutationId
          }
        }
    """

    inp = {"input": extract_serializer_input_fields(SaveAnswerSerializer, answer)}

    result = schema_executor(query, variable_values=inp)

    success = expected_error is None
    assert not bool(result.errors) == success
    if success:
        snapshot.assert_match(result.data)
    if not success:
        assert result.errors[0].extensions.get("code") == FORMAT_VALIDATION_FAILED
        assert result.errors[0].message == expected_error


@pytest.mark.parametrize(
    "question__type,question__format_validators",
    [(Question.TYPE_DATE, ["my-date-validator"])],
)
@pytest.mark.parametrize(
    "answer__value,success",
    [("2025-09-04", True), ("2025-09-05", False)],
)
def test_custom_format_validators(
    db, question, answer, success, schema_executor, settings
):
    settings.FORMAT_VALIDATOR_CLASSES = [
        "caluma.caluma_form.tests.test_format_validators.MyDateValidator",
    ]

    query = """
        mutation SaveDocumentDateAnswer($input: SaveDocumentDateAnswerInput!) {
          saveDocumentDateAnswer(input: $input) {
            clientMutationId
          }
        }
    """

    result = schema_executor(
        query,
        variable_values={
            "input": extract_serializer_input_fields(SaveAnswerSerializer, answer)
        },
    )
    assert not bool(result.errors) == success
    if not success:
        assert result.errors[0].extensions.get("code") == FORMAT_VALIDATION_FAILED
        assert result.errors[0].message == "5 is not an even day"
