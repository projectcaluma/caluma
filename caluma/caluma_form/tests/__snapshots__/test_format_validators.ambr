# name: test_base_format_validators[question__format_validators1-test@example.com-True-text]
  <class 'dict'> {
    'saveDocumentStringAnswer': <class 'dict'> {
      'answer': <class 'dict'> {
        'stringValue': 'test@example.com',
      },
      'clientMutationId': 'testid',
    },
  }
---
# name: test_base_format_validators[question__format_validators1-test@example.com-True-textarea]
  <class 'dict'> {
    'saveDocumentStringAnswer': <class 'dict'> {
      'answer': <class 'dict'> {
        'stringValue': 'test@example.com',
      },
      'clientMutationId': 'testid',
    },
  }
---
# name: test_base_format_validators[question__format_validators3-+411234567890-True-text]
  <class 'dict'> {
    'saveDocumentStringAnswer': <class 'dict'> {
      'answer': <class 'dict'> {
        'stringValue': '+411234567890',
      },
      'clientMutationId': 'testid',
    },
  }
---
# name: test_base_format_validators[question__format_validators3-+411234567890-True-textarea]
  <class 'dict'> {
    'saveDocumentStringAnswer': <class 'dict'> {
      'answer': <class 'dict'> {
        'stringValue': '+411234567890',
      },
      'clientMutationId': 'testid',
    },
  }
---
# name: test_fetch_format_validators
  <class 'dict'> {
    'allFormatValidators': <class 'dict'> {
      'edges': <class 'list'> [
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Not valid',
            'name': 'test name english',
            'regex': '(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+$)',
            'slug': 'test-validator',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Please enter a valid e-mail address',
            'name': 'E-mail address',
            'regex': '(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+$)',
            'slug': 'email',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Please enter a valid phone number',
            'name': 'Phone number',
            'regex': '^[\\s\\/\\.\\(\\)-]*(?:\\+|0|00)(?:[\\s\\/\\.\\(\\)-]*\\d[\\s\\/\\.\\(\\)-]*){6,20}$',
            'slug': 'phone-number',
          },
        },
      ],
      'pageInfo': <class 'dict'> {
        'endCursor': 'YXJyYXljb25uZWN0aW9uOjI=',
        'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
      },
      'totalCount': 3,
    },
  }
---
# name: test_fetch_format_validators.1
  <class 'dict'> {
    'allFormatValidators': <class 'dict'> {
      'edges': <class 'list'> [
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Nicht valid',
            'name': 'test name deutsch',
            'regex': '(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+$)',
            'slug': 'test-validator',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Bitte geben Sie eine gültige E-Mail Adresse ein',
            'name': 'E-Mail Adresse',
            'regex': '(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+$)',
            'slug': 'email',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Bitte geben Sie eine gültige Telefonnummer ein',
            'name': 'Telefonnummer',
            'regex': '^[\\s\\/\\.\\(\\)-]*(?:\\+|0|00)(?:[\\s\\/\\.\\(\\)-]*\\d[\\s\\/\\.\\(\\)-]*){6,20}$',
            'slug': 'phone-number',
          },
        },
      ],
      'pageInfo': <class 'dict'> {
        'endCursor': 'YXJyYXljb25uZWN0aW9uOjI=',
        'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
      },
      'totalCount': 3,
    },
  }
---
# name: test_fetch_format_validators.2
  <class 'dict'> {
    'allFormatValidators': <class 'dict'> {
      'edges': <class 'list'> [
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Not valid',
            'name': 'test name english',
            'regex': '(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+$)',
            'slug': 'test-validator',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Veuillez entrer une addresse e-mail valide',
            'name': 'Adresse e-mail',
            'regex': '(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+$)',
            'slug': 'email',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Veuillez entrer un numéro de téléphone valide',
            'name': 'Numéro de téléphone',
            'regex': '^[\\s\\/\\.\\(\\)-]*(?:\\+|0|00)(?:[\\s\\/\\.\\(\\)-]*\\d[\\s\\/\\.\\(\\)-]*){6,20}$',
            'slug': 'phone-number',
          },
        },
      ],
      'pageInfo': <class 'dict'> {
        'endCursor': 'YXJyYXljb25uZWN0aW9uOjI=',
        'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
      },
      'totalCount': 3,
    },
  }
---
# name: test_fetch_format_validators.3
  <class 'dict'> {
    'allFormatValidators': <class 'dict'> {
      'edges': <class 'list'> [
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Not valid',
            'name': 'test name english',
            'regex': '(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+$)',
            'slug': 'test-validator',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Please enter a valid e-mail address',
            'name': 'E-mail address',
            'regex': '(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+$)',
            'slug': 'email',
          },
        },
        <class 'dict'> {
          'node': <class 'dict'> {
            'errorMsg': 'Please enter a valid phone number',
            'name': 'Phone number',
            'regex': '^[\\s\\/\\.\\(\\)-]*(?:\\+|0|00)(?:[\\s\\/\\.\\(\\)-]*\\d[\\s\\/\\.\\(\\)-]*){6,20}$',
            'slug': 'phone-number',
          },
        },
      ],
      'pageInfo': <class 'dict'> {
        'endCursor': 'YXJyYXljb25uZWN0aW9uOjI=',
        'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
      },
      'totalCount': 3,
    },
  }
---
